### PR TITLE
ORV2-809: Changes for Client Number Generation And Application Status Update

### PR DESCRIPTION
**Pull request by gchauhan-aot**
# Description

This PR is for Client Number Generation and also update application status, handles special status updates like APPROVED, AUTO_APPROVED and ISSUED.
Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A Unit testing using postman
- [ ] Test B

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Further comments

---

Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend - Vehicles](https://onroutebc-489-backend-vehicles.apps.silver.devops.gov.bc.ca/) available
[Backend - DMS](https://onroutebc-489-backend-dms.apps.silver.devops.gov.bc.ca/) available
[Frontend](https://onroutebc-489-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge-main.yml)

---

Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend - Vehicles](https://onroutebc-489-backend-vehicles.apps.silver.devops.gov.bc.ca/) available
[Backend - DMS](https://onroutebc-489-backend-dms.apps.silver.devops.gov.bc.ca/) available
[Frontend](https://onroutebc-489-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge-main.yml)

---

Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend - Vehicles](https://onroutebc-489-backend-vehicles.apps.silver.devops.gov.bc.ca/) available
[Backend - DMS](https://onroutebc-489-backend-dms.apps.silver.devops.gov.bc.ca/) available
[Frontend](https://onroutebc-489-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge-main.yml)

### CODE REVIEW

File: backend/vehicles/src/modules/auth/jwt.strategy.ts [hunk 1]
```diff
@@ -48,6 +48,8 @@ export class JwtStrategy extends PassportStrategy(Strategy) {
       companyId = +req.params['companyId'];
     } else if (req.query['companyId']) {
       companyId = +req.query['companyId'];
+    } else if (req.body.companyId) {
```
Comment (praju-aot) [inline backend/vehicles/src/modules/auth/jwt.strategy.ts:53]: Kindly address the code lint errors. 

  51:16  error  Unsafe member access .companyId on an `any` value  @typescript-eslint/no-unsafe-member-access
  52:7   error  Unsafe assignment of an `any` value                @typescript-eslint/no-unsafe-assignment
  52:19  error  Unsafe member access. companyId on an `any` value  @typescript-eslint/no-unsafe-member-access

File: backend/vehicles/src/common/interface/code.interface.ts
Comment (praju-aot) [inline backend/vehicles/src/common/interface/code.interface.ts:1]: The file is to removed as its not being used.

File: backend/vehicles/src/modules/company-user-management/company/company.module.ts [hunk 2]
```diff
@@ -5,11 +5,12 @@ import { CompanyService } from './company.service';
 import { Company } from './entities/company.entity';
 import { CompanyProfile } from './profiles/company.profile';
 import { DataSource } from 'typeorm';
+import { DatabaseHelper } from 'src/common/helper/database.helper';
 
 @Module({
   imports: [TypeOrmModule.forFeature([Company])],
   controllers: [CompanyController],
-  providers: [CompanyService, CompanyProfile],
+  providers: [CompanyService, CompanyProfile, DatabaseHelper],
```
Comment (praju-aot) [inline backend/vehicles/src/modules/company-user-management/company/company.module.ts]: DatabaseHelper is registered as a provider to both Company and Application module. This needs to be revisited.

File: backend/vehicles/src/modules/company-user-management/company/company.service.ts [hunk 3]
```diff
@@ -264,4 +274,33 @@ export class CompanyService {
...
+  /**
+   * Generates clientNumber for the newly created company.
+   * @param company
+   *
+   */
+  private async generateClientNumber(
+    company: Company,
+    currentUser: IUserJWT,
+  ): Promise<string> {
+    const { randomInt } = await import('crypto');
```
Comment (praju-aot) [inline backend/vehicles/src/modules/company-user-management/company/company.service.ts]: Please import it at class level. 

import { randomInt } from 'crypto';

File: backend/vehicles/src/modules/company-user-management/company/company.service.ts [hunk 4]
```diff
@@ -65,6 +69,12 @@ export class CompanyService {
         },
       );
 
+      if (!newCompany.clientNumber) {
```
Comment (praju-aot) [inline backend/vehicles/src/modules/company-user-management/company/company.service.ts]: Please remove the check as the condition will always be true. The create method can only be called at the time of new company profile registration and hence client number will always be undefined/null.

File: backend/vehicles/src/modules/company-user-management/company/company.service.ts [hunk 5]
```diff
@@ -264,4 +274,33 @@ export class CompanyService {
...
+    company: Company,
+    currentUser: IUserJWT,
+  ): Promise<string> {
+    const { randomInt } = await import('crypto');
+    const rnd = randomInt(100, 1000);
+    const seq = await this.databaseHelper.callDatabaseSequence(
+      'dbo.ORBC_CLIENT_NUMBER_SEQ',
+    );
+    const accountSource =
+      currentUser.identity_provider == IDP.IDIR
```
Comment (praju-aot) [inline backend/vehicles/src/modules/company-user-management/company/company.service.ts]: The logic to moved to auth helper and value to be stored to payload.

File: backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts [hunk 6]
```diff
@@ -4,7 +4,7 @@ import { Base } from 'src/modules/common/entities/base.entity';
 import { Column, Entity, PrimaryColumn } from 'typeorm';
 
 @Entity({ name: 'permit.ORBC_VT_PERMIT_APPROVAL_SOURCE' })
-export class PermitApprovalSource extends Base {
+export class PermitApprovalSourceEntity extends Base {
   @AutoMap()
```
Comment (praju-aot) [inline backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts]: Please rename to PermitApprovalSource similar to naming conventions of other entities.

File: backend/vehicles/src/modules/permit/permit.module.ts [hunk 7]
```diff
@@ -11,8 +11,8 @@ import { ApplicationService } from './application.service';
 import { ApplicationProfile } from './profile/application.profile';
 import { PdfModule } from '../pdf/pdf.module';
 import { PermitType } from './entities/permit-type.entity';
-import { PermitApplicationOrigin } from './entities/permit-application-origin.entity';
-import { PermitApprovalSource } from './entities/permit-approval-source.entity';
+import { PermitApplicationOriginEntity } from './entities/permit-application-origin.entity';
```
Comment (praju-aot) [inline backend/vehicles/src/modules/permit/permit.module.ts]: Please rename to PermitApplicationOrigin similar to naming conventions of other entities.

File: backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { Base } from 'src/modules/common/entities/base.entity';
...
   @ApiProperty({
     example: 'ONLINE',
```
Comment (praju-aot) [inline backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts:11]: ID type should be of enum type PermitApprovalSource.

File: backend/vehicles/src/modules/permit/entities/permit-application-origin.entity.ts [hunk 9]
```diff
@@ -4,7 +4,7 @@ import { Base } from 'src/modules/common/entities/base.entity';
 import { Column, Entity, PrimaryColumn } from 'typeorm';
 
 @Entity({ name: 'permit.ORBC_VT_PERMIT_APPLICATION_ORIGIN' })
-export class PermitApplicationOrigin extends Base {
+export class PermitApplicationOriginEntity extends Base {
   @AutoMap()
   @ApiProperty({
     example: 'ONLINE',
```
Comment (praju-aot) [inline backend/vehicles/src/modules/permit/entities/permit-application-origin.entity.ts:11]: ID type should be of enum type PermitApplicationOrigin.

File: backend/vehicles/src/modules/auth/jwt.strategy.ts (hunk 1 above)
Comment (praju-aot) [reply backend/vehicles/src/modules/auth/jwt.strategy.ts:53]: Disabled the lint for the line as a temporary measure.

File: backend/vehicles/src/modules/permit/entities/permit-application-origin.entity.ts (hunk 9 above)
Comment (praju-aot) [reply backend/vehicles/src/modules/permit/entities/permit-application-origin.entity.ts:11]: Done

File: backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts (hunk 8 above)
Comment (praju-aot) [reply backend/vehicles/src/modules/permit/entities/permit-approval-source.entity.ts:11]: Done

**Review by praju-aot — APPROVED** [review]
+1

File: backend/vehicles/src/modules/auth/jwt.strategy.ts (hunk 1 above)
Comment (gchauhan-aot) [reply backend/vehicles/src/modules/auth/jwt.strategy.ts:53]: Done

File: backend/vehicles/src/common/interface/code.interface.ts
Comment (gchauhan-aot) [reply backend/vehicles/src/common/interface/code.interface.ts:1]: Done

**Pull request merged**